### ISSUE TITLE
Trackable effects on symptoms

### ISSUE THREAD

**Issue by lmerriam** [rubyforgood/Flaredown#145]
The ultimate goal of Flaredown as a tool for personal health is to help users quantify the effects of treatments and environmental stressors on their individual symptoms and condition as a whole.

A simple way to infer treatment effect would be to look at the average severity of a particular symptom on days when the treatment was taken, vs days when the treatment was not taken, and take the difference.

However, many medications are not intended to have same day effects. Some popular treatments are taken months apart and have a very long half life. 

Could we sample symptom severity multiple times after a treatment is taken? This could conflict with treatments that are taken every day, because we don't know whether an effect is coming from the prior day(s) or the current one.

Could we look at the time between days when a treatment is taken and infer the frequency at which the user is taking it, and base our sampling frequency on that?

Bottom line: how can we find the magnitude of a treatment or tags impact on a symptom, taking into account the fact that treatments/tag effects vary greatly in terms of the time it takes for their effect to occur.

Here are a couple brainstorming ideas about what we might want to show to the user, could be entirely different depending on how we calculate effects:
![image](https://cloud.githubusercontent.com/assets/1657444/21584421/8cc9e234-d075-11e6-9d7a-47606197034a.png)

**Comment by Graydyn** [comment]
I've pushed a script up to my github, https://github.com/Graydyn/flaredown_data/blob/master/treatment_effectiveness.py

Take a look, and let me know if there are any questions.  It will be up to the web team to load db fields into pandas, as I was working from the CSVs.

It returns a value from -5 to 5, with -5 meaning the user reports a 0 symptom rating when not on the treatment and 5 when on it, and 5 meaning the user reports a 5 symptom rating when off the treatment, and a 0 when on it.

There are two ways to measure the effectiveness.  You can look at the difference between days when the treatment was taken and not, and set whatever time offset you want (so you can say, treatment X makes you feel better about condition Y after about 2 days).  

Or you can look at the whole duration that a user is using a treatment, and compare to before and after that duration.  This method gives a large body of data to work with and tends to give much more confident results, but you should keep in mind that this method assumes that a user is taking a treatment regularly, and should only be used in cases where this assumption can be made.

To help select between the two methods, I've also provided a method to measure the periodicity of a treatment, which can be used to estimate if it's safe to use the windowing method.

**Comment by lmerriam** [comment]
Pulling in the information from #151 so I can close it, because these issues are too similar:

<img width="548" alt="screenshot 2017-01-10 14 59 54" src="https://cloud.githubusercontent.com/assets/1657444/21869830/7042bffa-d827-11e6-9eac-14601e3dc1a2.png">

I propose that we create a page for each `trackable_type` that shows these correlations both for the individual user and for our entire population. The above is an example of a `condition` page. We can use the formula provided by Graydyn in #145 (working on an update) to do the analysis on that page. For the "only me" section I believe this requires us to correlate each symptom/condition with every other trackable the user is tracking, and then aggregate all of those individual correlations to find the numbers for the "everyone" page. @Graydyn  would we do the aggregate correlations in the same way that we do the individual user correlations?

**Comment by lmerriam** [comment]
@Graydyn after talking with @mpugach we need to figure out the specifics of how to hook Flaredown data into these python scripts. The situation we want to avoid is one where each time we update the script Max has to manually go in and edit it to make it connect to the DB. Ideally we could get it hooked in and then future updates could be pushed directly.

This is getting a little over my head at this point, so I'm hoping one of you guys has an idea about how we can create a good workflow.

**Comment by mpugach** [comment]
It seems readable enough to be translated in Ruby, but it will be pain to port Python logic changes every time.

I'm not very familiar with the language, so will not be effective to write the DB logic. But if no one can do it, then why not to learn some Python? :)

**Comment by Graydyn** [comment]
I think porting to Ruby would be hard because of the Pandas dataframes.  Unless Ruby has something similar?
I think the effectiveness code could be rewritten in Ruby without dataframes, but we might find ourselves in a similar situation in the future, when we want to start running stats code that does need dataframes.

I'd recommend either writing the DB code in Python, or creating a Ruby interface that I can hit with Python code to get data.  I haven't looked at infrastructure at all (been purposely keeping my nose out of it).  If you guys can run microservices, then having a Ruby API that feeds data to a Python API would work nicely.

**Comment by mpugach** [comment]
We can store prepared data in a separate MongoDB table(collection). By preparations I mean - no joins and calculations needed, because they will be done on every checkin insert/update in async background job. This is on Ruby side.

So the Python script will connect directly to Mongo without Puma(webserver) in the middle. It will be much faster, since there is a lot off data. And it will not waste the web worker's connections.

Also Mongo aggregation features can be used if needed. We can also do this on preparation step, to not force the Python developer get too much familiar with Mongo.

After the data is loaded and processed, the results can be stored in a separate Mongo collection. There should be some kind of iteration identifier in every document. Simple increment should be fine.

**Comment by lmerriam** [comment]
@Graydyn last time we talked you said you had an idea for an improved algorithm (I think it was after we looked at this paper on financial analysis). Super curious about what you're thinking!

@espowell here's where we're doing most of the discussion about trackable (ie symptoms, treatments, food, weather, etc that people log in FD) effects on symptoms. I believe that you could substitute tags for food until we have a meaningful amount of food data, as they are logged in exactly the same way.

Here's a list of the problems we have identified so far, let me know if there are other considerations:
- Missing data: we don't know that a user didn't encounter a food just because they didn't log it.
- Treatments taken as-needed: some treatments are taken specifically to halt or reverse a flare. Because they are taken when symptoms are already worsening, they may appear to be negatively affecting the user even if they are actually effective.
- Lag of effect: some treatments can take effect on the same day, and some have a half life of months.

(I used "treatments" for brevity, but most of the above applies to other types of trackable as well)

**Comment by lmerriam** [comment]
@mpugach if we forget about setting up a python environment for this first version, can you estimate a very simple version of this based on this formula:

(average symptom activity on days when the treatment is taken) - (average symptom activity on days when the treatment is not taken) = effect of treatment

We would run this on every combination of condition/symptom vs every other trackable type. Depending on processing power required, we could also run this on multiple dates after a trackable is indicated on the check-in to estimate its effect over time.

**Comment by mpugach** [comment]
> (average symptom activity on days when the treatment is taken) - (average symptom activity on days when the treatment is not taken) = effect of treatment
> We would run this on every combination of condition/symptom vs every other trackable type.

How should we compute the effect of weather?
We know when the user does not take his treatment, because he is tracking it. But what about food? 

> Depending on processing power required, we could also run this on multiple dates after a trackable is indicated on the check-in to estimate its effect over time.

Can you provide an example please?

**Comment by lmerriam** [comment]
> How should we compute the effect of weather?

Will get back to you on this, but I talked with graydyn about it and it's simple correlation math. However, maybe we should start with just treatments/tags/food for now?

> We know when the user does not take his treatment, because he is tracking it. But what about food?

Good question. For now this is a dumb algorithm that assumes the logging is complete and we will make that clear to the user so they can interpret the data appropriately. An upgraded version of this algorithm (mentioned in a comment above above) will try to account for missing data and other problems that we see with this approach.

> Can you provide an example please?

We could run the same equation for average symptom activity _two days after_ a treatment was taken, for example. In the mock at the very top of this issue there's "same day, 3 days, 1 week" which might be a decent timeframe.

**Comment by espowell** [comment]
For tracking food- It seems like the easiest route right now is to use tags to indicate nutrients of interest. Most people who are tracking diet for chronic diseases are concerned about specific nutrients or items- dairy, caffeine, red meat, etc- and know when they've eaten it. If we want more complex ways of tracking diet I'm happy to discuss that! 

> Good question. For now this is a dumb algorithm that assumes the logging is complete and we will make that clear to the user so they can interpret the data appropriately. An upgraded version of this algorithm (mentioned in a comment above above) will try to account for missing data and other problems that we see with this approach.

For both food, and other things logged- assuming logging is complete can be problematic. I know we need to make some assumptions for simplicity sake, but would it be too complicated for users to log when they didn't do/take something? For example logging "didn't eat dairy" would be a 0 and "ate dairy" would be a 1 and days without the tag logged would be missing. 

Alternatively, a reminder could reduce missing data. Would it be possible to develop a notification of some kind? Something that pops up/emails at the end of the day and says, Hey- it looks like you didn't log this tag. Did you encounter this today? 

> We could run the same equation for average symptom activity two days after a treatment was taken, for example. 

One issue with the simple algorithm is that it doesn't account for other treatments taken over the same time period. So say you ate dairy + caffeine + bad weather. The simple algorithm says each one of those three things completely accounted for the change in symptoms that day. This is a particularly big problem for things which tend to occur together (like caffeine+ a poor nights sleep).

I know we've got to keep it simple for the sake of processing power- would it take way too much to do one linear regression with all the tags in it? That would help tease out the simultaneous treatment issue.

**Comment by Graydyn** [comment]
Personally I don't see a reason to treat food or weather as different from any other tag.  If people want to tag certain nutrients than they are free to do so.  Many of them already track gluten.

Linear regression is computational cheap but I'm not clear on what you're trying to accomplish with it.
About the confounding issue, if a user always uses the same treatments on the same day, certainty about whether those treatments are working for that specific user will never be attained.  P values will never exceed %50.  No algorithm can fix this problem, the information just isn't there.
You could supply supplemental information to the user by looking at the larger population.  You can say "we don't know if it's treatment X or treatment Y making you feel better, but in general people like you tend to feel better when they take treatment X".  If you want to get into that, go ahead and read recommender_system.ipynb in my github.

**Comment by lmerriam** [comment]
Awesome points being made. We're currently discussing this as if it were one component of Flaredown, but I believe we can actually split it into two different flows:

For the _daily checkin_ stuff, I don't know that we'll ever get to super confidence on tags/foods. That's necessitated by the fact that tags/food can be different every day -- we can't keep them on users' check-ins forever because it will quickly become a massive list. We might be able to improve it by being smarter about auto-suggestion or keeping "recent tags" etc around longer though.

But we can get more serious, validated measures when we implement PROMs and other bonus questions like food surveys (very soon! maybe this week!) -- because at that point we can do exactly what @espowell is suggesting and ask users targeted questions. I think that makes perfect sense for questions like the food categories. Does that make sense?

Should I start a thread where we can think about the longer-term, more comprehensive algorithm for analyzing correlations?

**Comment by espowell** [comment]
> Should I start a thread where we can think about the longer-term, more comprehensive algorithm for analyzing correlations?

I think that's a great idea! That way we won't clog up threads. 

@Graydyn I'll definitely check out the recommender_system - Sounds like a great add, especially for newer users who might not have a lot of data logged.

>  In the mock at the very top of this issue there's "same day, 3 days, 1 week" which might be a decent timeframe.

The more we talk this through, the more I realize that calculating over a longer time frame would be helpful for getting more accurate estimates. Could we add a 1 month or 3 months to this? Unless I misunderstood your comment @lmerriam in which case ignore this :)

> But we can get more serious, validated measures when we implement PROMs and other bonus questions like food surveys (very soon! maybe this week!) 

Sweet! What kind of food surveys are you thinking? 

My thought with the linear regression is that it could better account for the fact that simultaneous treatments were being logged. I was worried that the simple approach would overestimate effects because it didn't account for other treatments being logged. But regression definitely doesn't help with the collinearity problem- so perhaps its a wash.

**Comment by compwron** [comment]
Closing as old for now

**Comment by compwron** [comment]
@bklang this one could be very interesting for a deeper dive from design
